src/System.Device.Gpio/System/Device/Board.cs

```diff
@@ -0,0 +1,246 @@
+            {
+                if (_pinReservations.TryGetValue(logicalPin, out var reservation))
+                {
+                    if (reservation.Owner != owner || reservation.Usage != usage)
+                    {
+                        throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not the owner or the usage is wrong. Class {reservation.Owner} has reserved the Pin for {reservation.Usage}");
+                    }
+
+                    _pinReservations.Remove(logicalPin);
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because it is not reserved.");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Override this method if something special needs to be done to use the pin for the given device.
+        /// Many devices support multiple functions per Pin, but not at the same time, so that some kind of
+        /// multiplexer needs to be set accordingly.
+        /// </summary>
+        /// <param name="pinNumber">The logical pin number to use.</param>
+        /// <param name="usage">The intended usage</param>
+        protected virtual void ActivatePinMode(int pinNumber, PinUsage usage)
+        {
+        }
+
+        public virtual void Initialize()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(ToString());
+            }
+
+            _initialized = true;
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            _disposed = true;
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private sealed class PinReservation
+        {
+            public PinReservation(int pin, PinUsage usage, object owner)
+            {
+                Pin = pin;
+                Usage = usage;
+                Owner = owner;
+            }
+
+            public int Pin { get; }
+            public PinUsage Usage { get; }
+
+            /// <summary>
+            /// Component that owns the pin (used mainly for debugging)
+            /// </summary>
+            public object Owner { get; }
+        }
+
+        public abstract GpioController CreateGpioController(PinNumberingScheme pinNumberingScheme);
+        public abstract I2cDevice CreateI2cDevice(I2cConnectionSettings connectionSettings);
+
+        public abstract SpiDevice CreateSpiDevice(SpiConnectionSettings settings);
+
+        public abstract PwmChannel CreatePwmChannel(
+            int chip,
+            int channel,
+            int frequency = 400,
+            double dutyCyclePercentage = 0.5);
+
+        public abstract AnalogController CreateAnalogController(int chip);
+
+        public static Board DetermineOptimalBoardForHardware(PinNumberingScheme defaultNumberingScheme = PinNumberingScheme.Logical)
```

Comment:
This should only be used for implementation detail, shouldn't be exposed

Reply:
This is the factory method you suggested to use, so making it internal is not really helpful, I suppose.

Reply:
I see, I didn't realize it was returning a Board. In that case the naming is confusing, factory methods are usually called something like Create()